src/webgpu/api/operation/queue/writeBuffer.spec.ts

```diff
@@ -1,11 +1,228 @@
-export const description = `
-TODO:
-- source.origin is unaligned
-- data given as ArrayBuffer
-- data given as each TypedArray variant
-`;
+export const description = 'Operation tests for GPUQueue.writeBuffer()';
 
+import { params, pbool, poptions } from '../../../../common/framework/params_builder.js';
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { range } from '../../../../common/framework/util/util.js';
 import { GPUTest } from '../../../gpu_test.js';
 
-export const g = makeTestGroup(GPUTest);
+const kTypedArrays = [
+  'Uint8Array',
+  'Uint16Array',
+  'Uint32Array',
+  'Int8Array',
+  'Int16Array',
+  'Int32Array',
+  'Float32Array',
+  'Float64Array',
+] as const;
+
+type WriteBufferSignature = {
+  bufferOffset: number;
+  data: readonly number[];
+  arrayType: typeof kTypedArrays[number];
+  useArrayBuffer: boolean;
+  dataOffset?: number;
+  dataSize?: number;
+};
+
+class F extends GPUTest {
+  calculateRequiredBufferSize(writes: WriteBufferSignature[]): number {
+    let bufferSize = 0;
+    // Calculate size of final buffer
+    for (const { bufferOffset, data, arrayType, useArrayBuffer, dataOffset, dataSize } of writes) {
+      const typeBuilder = self[arrayType];
+      let requiredSize = data.length * typeBuilder.BYTES_PER_ELEMENT;
+
+      // When passing data as an ArrayBuffer, dataOffset, dataSize use byte instead of number of
+      // elements
+      const bytesPerElement = useArrayBuffer ? 1 : typeBuilder.BYTES_PER_ELEMENT;
+      if (dataOffset) {
+        requiredSize -= dataOffset * bytesPerElement;
+      }
+      if (dataSize) {
+        requiredSize = Math.min(requiredSize, dataSize * bytesPerElement);
+      }
+
+      requiredSize += bufferOffset;
+
+      if (requiredSize > bufferSize) {
+        bufferSize = requiredSize;
+      }
+    }
+    // writeBuffer requires buffers to be a multiple of 4
+    return Math.ceil(bufferSize / 4) * 4;
+  }
+
+  testWriteBuffer(...writes: WriteBufferSignature[]) {
+    const bufferSize = this.calculateRequiredBufferSize(writes);
+    const buffer = this.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    // Initialize buffer to non-zero data (0xff)
```

Comment:
Uh, this is using writeBuffer to test writeBuffer. Maybe use mappedAtCreation instead?

Reply:
I think this is okay; if this writeBuffer fails, then the expectContents will still fail.